rust/flatc/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+//!
+//! Include the generated Rust code:
+//!
+//! ```ignore
+//! mod foo_generated {
+//!     include!(concat!(env!("OUT_DIR"), "/foo_generated.rs"));
+//! }
+//! ```
+
+#![allow(clippy::new_without_default_derive)]
+
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::{env, fmt, io};
+
+/// An internal error that occured.
+struct Error {
+    /// The kind of error.
+    kind: ErrorKind,
+    /// A helpful message.
+    message: String,
+}
+
+impl Error {
+    /// Create a new instance of an `Error`.
+    fn new<S: AsRef<str>>(kind: ErrorKind, message: S) -> Self {
+        Self {
+            kind,
+            message: message.as_ref().to_owned(),
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "An {} error occurred during FlatBuffer compilation: {}",
+            self.kind, self.message
+        )
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(error: io::Error) -> Self {
+        Self::new(ErrorKind::IO, format!("{}", error))
+    }
+}
+
+/// The type of errors that can occur when using `flatc`.
+enum ErrorKind {
+    /// An I/O error occured.
+    IO,
+    /// An error occurred while using the schema compiler.
+    FlatC,
+}
+
+impl fmt::Display for ErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            ErrorKind::IO => write!(f, "I/O"),
+            ErrorKind::FlatC => write!(f, "flatc internal"),
+        }
+    }
+}
+
+/// A builder for compilation of one or multiple FlatBuffer schemas.
+///
+/// This is the core type of `flatc`. For further documentation, see the
+/// individual methods for this structure.
+pub struct Build {
+    /// The path that compiled schema files will be written into.
+    output_path: Option<PathBuf>,
+    /// The paths that will be tried in `include` statements in schemas.
+    include_paths: Vec<PathBuf>,
+    /// The schema files that will be compiled.
+    schema: Vec<PathBuf>,
+}
+
+impl Build {
+    /// Create a new instance of a `flatc` compilation with no configuration.
+    ///
+    /// This is finished with the [`compile`](struct.Build.html#method.compile)
+    /// function.
+    pub fn new() -> Self {
+        Self {
+            output_path: None,
+            include_paths: Vec::new(),
+            schema: Vec::new(),
+        }
+    }
+
+    /// Add a schema file to the files which will be compiled.
+    ///
+    /// ## Shell Command
+    ///
+    /// ```text
+    /// $ flatc -r foo.fbs bar.fbs baz.fbs
+    /// ```
+    ///
+    /// ## Rust Equivalent
+    ///
+    /// ```no_run
+    /// extern crate flatc;
+    ///
+    /// fn main() {
+    ///     flatc::Build::new()
+    ///         .schema("foo.fbs")
+    ///         .schema("bar.fbs")
+    ///         .schema("baz.fbs")
+    ///         .compile();
+    /// }
+    /// ```
+    pub fn schema<P: AsRef<Path>>(&mut self, file: P) -> &mut Self {
```

Comment:
Wasn't sure what to name this argument as (adding FBS files). I think it might be better of as distinct from `--schema`.